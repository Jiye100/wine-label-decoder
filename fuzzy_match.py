from rapidfuzz import process, fuzz
import string

# using rapidfuzz instead of fuzzywuzzy because it's faster

# grape variety
grape_variety = ['pinotage',
                 'cabernet sauvignon',
                 'pinot noir',
                 'sauvignon blanc',
                 'malbec',
                 'riesling',
                 'chardonnay',
                 'carmenere',
                 'torrentes', 
                 'pinot gris', 
                 'syrah', 
                 'chenin blanc', 
                 'merlot',
                 'zinfandel',
                 'cabernet franc',
                 'viognier',
                 'shiraz',
                 'bonarda',
                 'semillon',
                 'muscat',
                 'muscadet',
                 'gewurztraminer',
                 'gamay',
                 'cabernet',
                 'sauvignon'
                ]

regions = ["columbia valley", "yakima valley", "walla walla valley", "red mountain", "horse heaven hills",
           "willamette valley", "dundee hills", "yamhill-carlton", "mcminnville", "ribbon ridge", "chehalem mountains", "umpqua valley", "rogue valley", "applegate valley",
           "finger lakes", "cayuga lake", "seneca lake", "north fork of long island", "hamptons", "hudson river region", "lake erie",
           "napa", "napa valley", 'sonoma valley', 'russian river valley', 'alexander valley', 'sonoma county','mendocino','santa barbara county','monterey county','paso robles',
           "mendoza", "san juan", "salta", "la rioja", "catamarca",
           "marlborough", "martinborough", "central otago",
           "western cape", "stellenbosch", "simonsberg",
           "elqui valley", "limari valley", "aconcagua", "maipo valley", "cachapoal valley", "colchagua valley", "maule valley",
           "hunter valley", "yarra valley", "adelaide hills", "barossa valley", "clare valley", "mclaren vale", "margaret river", "rutherglen", "coonawarra", "mornington peninsula", "tasmania",
           "nantais", "anjou", "savennieres", "saumur", "touraine", "sancerre", "pouilly-fume", "reuilly", "quincy", "vouvray", "chinon", "bourgueil"
           ]

country_to_region = {
    "california": ["napa valley", 'sonoma valley', 'russian river valley', 'alexander valley', 'sonoma county','mendocino','santa barbara county','monterey county','paso robles'],
    "washington": ["columbia valley", "yakima valley", "walla walla valley", "red mountain", "horse heaven hills"],
    "new york": ["finger lakes", "cayuga lake", "seneca lake", "north fork of long island", "hamptons", "hudson river region", "lake erie"],
    "oregon": ["willamette valley", "dundee hills", "yamhill-carlton", "mcminnville", "ribbon ridge", "chehalem mountains", "umpqua valley", "rogue valley", "applegate valley"],
    "argentina": ["mendoza", "san juan", "salta", "la rioja", "catamarca"],
    "new zealand": ["marlborough", "martinborough", "central otago"],
    "south africa": ["western cape", "stellenbosch", "simonsberg"],
    "chile": ["elqui valley", "limari valley", "aconcagua", "maipo valley", "cachapoal valley", "colchagua valley", "maule valley"],
    "australia": ["hunter valley", "yarra valley", "adelaide hills", "barossa valley", "clare valley", "mclaren vale", "margaret river", "rutherglen", "coonawarra", "mornington peninsula", "tasmania", "geographe"]
}

possible_keywords = ["estate", "e$tate", "reserva", "d.o", "d.0", "d,o", "d,0", "sur lie"
                    #  "vins de france", "igp", "i.g.p", "indication geographique protegee", "aoc", "appellation d'origine controlee".
                    #  "cremant", "saumur brut", "montlouis sur loire", "anjou mousseux", "touraine mousseux", "vouvray", "grand cru",
                    #  "vendanges tardives", "selection de grains nobles", "gentil", "edelzwicker"
                     ]
translator = str.maketrans('', '', string.punctuation)
keyword = {"estate": "estate", 
           "e$tate": "estate",
           "reserve": "reserve",
           "reserva": "reserve",
           "d.o": "d.o.",
           "d.0": "d.o.",
           "d,o": "d.o.",
           "d,0": "d.o.",
           "sur lie": "sur lie"
           }

region_to_country = {region: country for country in country_to_region.keys() for region in country_to_region[country]}

# country or states
country_state = ["california", "oregon", "new york", "washington", "south africa", "argentina", "new zealand", "chile", "australia", "loire", "alsace"]
state = ['california', 'oregon', 'new york', 'washington']

# common error words that should be ignored
common_error = ['vineyard', 'region', 'valley', 'hills', 'vineyards', 'hill']

# fuzzy matching for grape varieties
def fuzzy_matching_grapes(result, country=None):
    best_match = None
    matching_text = None
    best_score = 0 # no threshold

    cabernet, sauvignon = False, False

    for text in result:
        text = text.lower()

        # ignore common errors or short text
        if text in common_error or len(text) < 4:
            continue

        # find the best match
        choice, score, _ = process.extractOne(text, grape_variety)

        if choice == 'cabernet' and score > 85:
            cabernet = True
        elif choice == 'sauvignon':
            sauvignon = True
        elif best_score < score:
            best_match = choice
            best_score = score
            matching_text = text

        if cabernet and sauvignon:
            best_match = 'cabernet sauvignon' 
            break
        elif sauvignon:
            best_match = 'sauvignon blanc'
            break
        elif cabernet:
            best_match = 'cabernet franc'
            break
        
    #     # torrentes is argentina's signature grape (unlikely to be seen if region is not argentina)
    #     if country != 'argentina' and choice == 'torrentes':
    #         continue

    #     # pinotage is south africa's signature grape (unlikely to be seen if region is not south africa)
    #     if country != 'south africa' and choice == 'pinotage':
    #         continue

    #     # compute Levenshtein Distance similarity ratio between two strings
    #     ratio_score = fuzz.ratio(choice, text)
    #     token_set_score = fuzz.token_set_ratio(choice, text)

    #     if best_score < token_set_score:
    #         best_match = choice
    #         best_score = token_set_score
    #         matching_text = text
    #     elif best_score < ratio_score:
    #         best_match = choice
    #         best_score = ratio_score
    #         matching_text = text


    return best_match, best_score, matching_text

def fuzzy_matching_regions(result):
    best_match = None
    best_score = 60 # threshold of 60
    matching_text = None

    for text in result:
        text = text.lower()

        # ignore common errors or grape varieties or short text
        if text in common_error or text in grape_variety or len(text) < 4:
            continue

        # find best match using token set ratio and partial ratio
        token_choice, token_score, _ = process.extractOne(text, regions, scorer=fuzz.token_set_ratio)
        partial_choice, partial_score, _ = process.extractOne(text.replace(' ', ''), regions, scorer=fuzz.partial_ratio)
        choice, score = None, 0
        # print("text:", text)
        # print("token scoring:", token_choice, token_score)
        # print("partial scoring:", partial_choice, partial_score, "\n")
        if token_score < partial_score:
            choice, score = partial_choice, partial_score
        else:
            choice, score = token_choice, token_score

        # short tokens need to have higher threshold
        if choice == 'salta':
            if score > 90:
                best_match = choice
                best_score = score
                matching_text = text
            continue

        if choice == 'napa':
            if score == 100:
                best_match = 'napa valley'
                best_score = score
                matching_text = text
                break

        if best_score < score:
            best_match = choice
            best_score = score
            matching_text = text


    return best_match, best_score, matching_text

def fuzzy_matching_country(result):
    best_match = None
    best_score = 60 # threshold of 60
    matching_text = None

    for text in result:
        text = text.lower()

        # ignore common errors or grape varieties or short text
        if text in common_error or text in grape_variety or len(text) < 4:
            continue

        # find best match
        choice, _, _ = process.extractOne(text, country_state)

        # compute Levenshtein Distance similarity ratio between two strings
        ratio_score = fuzz.ratio(choice, text)

        # OCR can read countries/states with high accuracy
        partial_token_set_score = fuzz.token_set_ratio(choice, text)

        # check if Levenshtein Distance similarity ratio between two strings is higher
        if partial_token_set_score == 100:
            best_match = choice
            best_score = partial_token_set_score
            matching_text = text
            break
        elif best_score < ratio_score:
            best_match = choice
            best_score = ratio_score
            matching_text = text
    
    return best_match, best_score, matching_text

def fuzzy_matching_keyword(result):
    scores = set()
    threshold = 85
    for text in result:
        text = text.lower()

        # ignore short words
        if len(text) < 4:
            continue

        token_choice, token_score, _ = process.extractOne(text, possible_keywords, scorer=fuzz.partial_token_set_ratio)
        partial_choice, partial_score, _ = process.extractOne(text.replace(' ', ''), possible_keywords, scorer=fuzz.partial_ratio)

        if token_score < partial_score:
            choice, score = keyword[partial_choice], partial_score
        else:
            choice, score = keyword[token_choice], token_score

        # print("text", text)
        # print("choice", choice, f"({score})")

        if choice == 'd.o.':
            # only add D.O. to keyword when it's 100% certain (tends to get higher similarity score because it's short)
            if score == 100:
                scores.add(choice)
            continue

        # ignore if similarity score is high because of the word 'state'
        if choice == 'estate' and score < fuzz.partial_ratio(text.lower(), 'state'):
            continue

        if score > threshold:
            scores.add(choice)

    return scores

def extract_info(result):
    best_country, country_score, country_matching  = fuzzy_matching_country(result)
    best_region, region_score, region_matching = fuzzy_matching_regions(result)

    # if country is not found or us wine (need to find AVA)
    if best_country is None or best_country in state:
        # check if state wine (AVA does not match state)
        if best_country in state and best_region not in country_to_region[best_country]:
            best_region = None
        else:
            # find country using region
            best_country = region_to_country[best_region]

    # check if region and country match (set it to match the higher scored one)
    if best_region is not None and best_region not in country_to_region[best_country]:
        if region_score < country_score:
            best_region = None
        else:
            best_country = region_to_country[best_region]

    # find grape variety
    best_grape, _, _ = fuzzy_matching_grapes(result, best_country)

    # find keywords
    keywords = fuzzy_matching_keyword(result)

    return best_grape, best_region, best_country, keywords