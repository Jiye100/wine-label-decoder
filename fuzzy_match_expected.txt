img1
grape: malbec
region: mendoza
country: argentina

img2
grape: viognier
region: mclaren vale
country: australia

img3
grape: cabernet sauvignon
region: napa valley
country: california

img4
grape: malbec
region: mendoza
country: argentina

img5
grape: cabernet sauvignon
region: mendoza
country: argentina

img6
grape: chardonnay
region: napa valley
country: california

img7
grape: riesling
country: australia

img8
grape: zinfandel
region: paso robles
country: california

img9
grape: shiraz 
region: barossa valley
country: australia

img10
grape: cabernet sauvignon
country: california

img11
grape: cabernet sauvignon
region: stellenbosch
country: south africa

img12
grape: pinot noir
region: marlborough

img13
grape: cabernet sauvignon
region: napa valley

img14
grape: cabernet sauvignon
region: north fork of long island
country: new york

img15
grape: pinot noir
region: mendoza
country: argentina

img16
grape: chardonnay
region: russian river valley (sonoma county)

img17
grape: pinot gris
region: willamette valley
country: oregon

img18
grape: pinot noir
region: central otago
country: new zealand

img19
grape: chardonnay
region: margaret river
country: australia

img20
grape: sauvignon blanc
region: marlborough

img21
grape: chardonnay
region: russian river valley (sonoma county)

img22
grape: zinfandel
region: paso robles

img23
grape: zinfandel
country: california

img24
grape: cabernet sauvignon
region: napa valley

img25
grape: sauvignon blanc
region: napa valley

img26
grape: sauvignon blanc
region: alexander valley (sonoma county)

img27
grape: chardonnay
region: russian river valley (sonoma county)

img28
grape: malbec
country: argentina

img29
grape: malbec
region: mendoza
country: argentina

img30
grape: pinot noir
country: california

img31
grape: cabernet sauvignon
region: napa valley

img32
grape: torrentes
region: salta
country: argentina

img33
grape: sauvignon blanc
region: marlborough
country: new zealand

img34
grape: cabernet sauvignon
region: napa valley

img35
grape: pinot noir
country: california

img36
grape: merlot
region: napa valley

img37
grape: shiraz
region: adelaide hills

img38
grape: zinfandel
region: paso robles
country: california

img39
grape: riesling
region: monterey
country: california

img40
grape: cabernet sauvignon
region: mendocino county

img41
grape: cabernet sauvignon
region: napa valley

img42
grape: chardonnay
region: dundee hills

img43
grape: pinot gris
region: willamette valley

img44
grape: chardonnay
region: monterey county, santa barbara county

img45
grape: pinot noir
region: santa barbara county

img46
grape: malbec
region: san juan
country: argentina

img47
grape: cabernet sauvignon
region: columbia valley

img48
grape: riesling
country: washington

img49
grape: pinot noir
region: marlborough
country: new zealand

img50
grape: cabernet sauvignon
region: alexander valley

img51
grape: pinot noir
region: willamette valley

img52
grape: cabernet sauvignon
region: sonoma county

img53
grape: shiraz
country: australia

img54
grape: cabernet sauvignon
region: napa valley

img55
grape: chardonnay
region: napa valley

img56
grape: pinotage
region: stellenbosch
country: south africa

img57
grape: pinotage
region: western cape, stellenbosch
country: south africa

img58
grape: cabernet sauvignon
region: napa valley
country: california

img59
grape: pinot noir
region: central otago
country: new zealand

img60
grape: sauvignon blanc
region: marlborough
country: new zealand

img61
grape: sauvignon blanc
region: marlborough
country: new zealand

img62
grape: pinot noir
region: marlborough
country: new zealand

img63
grape: sauvignon blanc
region: marlborough
country: new zealand

img64
grape: pinot noir
region: willamette valley

img65
grape: pinot noir
region: willamette valley

img66
grape: malbec
region: mendoza
country: argentina

img67
grape: merlot
country: washington

img68
grape: malbec
region: mendoza
country: argentina

img69
grape: riesling
region: yakima valley

img70
grape: chardonnay
region: curico valley?
country: chile

img71
grape: carmenere
country: chile

img72
grape: torrentes
region: mendoza
country: argentina

img73
grape: pinot noir
region: willamette valley
country: oregon

img74
grape: chardonnay
region: columbia valley

img75
grape: merlot
region: columbia valley

img76
grape: cabernet sauvignon
region: walla walla valley
country: washington